be/src/storage/primary_index.cpp

```diff
@@ -1502,6 +1502,7 @@ Status PrimaryIndex::get(const Column& key_col, std::vector<uint64_t>* rowids) c
 }
 
 std::size_t PrimaryIndex::memory_usage() const {
+    std::lock_guard<std::mutex> lg(const_cast<std::mutex&>(_lock));
```

Comment:
Will this influence performance of this function? Because this lock will make the caller stuck to wait for the lock.

Reply:
> Will this influence performance of this function? Because this lock will make the caller stuck to wait for the lock.

The functions that hold this lock perform purely in-memory lightweight operations. The only function that involves IO is during the initialization of the PrimaryIndex, which does not involve any lock contention. Therefore, I believe this has minimal impact on performance.

Reply:
> > Will this influence performance of this function? Because this lock will make the caller stuck to wait for the lock.
> 
> 
> 
> The functions that hold this lock are all lightweight memory operations, so I believe the impact on performance is minimal. And I think your concern is reasonable, so I add atomic val `_memory_usage`, calculate it each time when primary index been modified, and `PrimaryIndex::memory_usage` is only need to return `_memory_usage`, which is thread-safe and no need to hold any lock.

That will be better. I don't worry about this function’s cost. What I worry is that other functions will hold this lock for a long time so that the caller of this function will stuck waiting for the lock. But you have removed the lock then my concern is gone.